File: java/test/behaviour/concept/type/attributetype/AttributeTypeSteps.java
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.vaticle.typedb.client.test.behaviour.concept.type.attributetype;

import com.vaticle.typedb.client.api.concept.Concept;
import com.vaticle.typedb.client.api.concept.type.AttributeType;
import com.vaticle.typedb.client.api.concept.type.ThingType.Annotation;
import com.vaticle.typedb.client.api.concept.value.Value;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

import java.util.List;
import java.util.Set;

import static com.vaticle.typedb.client.api.concept.Concept.Transitivity.EXPLICIT;
import static com.vaticle.typedb.client.test.behaviour.connection.ConnectionStepsBase.tx;
import static com.vaticle.typedb.common.collection.Collections.set;
import static java.util.Collections.emptySet;
import static java.util.stream.Collectors.toSet;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

public class AttributeTypeSteps {

    @When("put attribute type: {type_label}, with value type: {value_type}")
    public void put_attribute_type_with_value_type(String typeLabel, Value.Type valueType) {
        tx().concepts().putAttributeType(typeLabel, valueType);
    }

    @Then("attribute\\( ?{type_label} ?) get value type: {value_type}")
    public void attribute_type_get_value_type(String typeLabel, Value.Type valueType) {
        assertEquals(valueType, tx().concepts().getAttributeType(typeLabel).getValueType());
    }

    @Then("attribute\\( ?{type_label} ?) get supertype value type: {value_type}")
    public void attribute_type_get_supertype_value_type(String typeLabel, Value.Type valueType) {
        AttributeType supertype = tx().concepts().getAttributeType(typeLabel).getSupertype(tx()).asAttributeType();
        assertEquals(valueType, supertype.getValueType());
    }

    @Then("attribute\\( ?{type_label} ?) as\\( ?{value_type} ?) get subtypes contain:")
    public void attribute_type_as_value_type_get_subtypes_contain(String typeLabel, Value.Type valueType, List<String> subLabels) {
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        Set<String> actuals = attributeType.getSubtypes(tx(), valueType).map(t -> t.getLabel().name()).collect(toSet());
        assertTrue(actuals.containsAll(subLabels));
    }

    @Then("attribute\\( ?{type_label} ?) as\\( ?{value_type} ?) get subtypes do not contain:")
    public void attribute_type_as_value_type_get_subtypes_do_not_contain(String typeLabel, Value.Type valueType, List<String> subLabels) {
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        Set<String> actuals = attributeType.getSubtypes(tx(), valueType).map(t -> t.getLabel().name()).collect(toSet());
        for (String subLabel : subLabels) {
            assertFalse(actuals.contains(subLabel));
        }
    }

    @Then("attribute\\( ?{type_label} ?) as\\( ?{value_type} ?) set regex: {}")
    public void attribute_type_as_value_type_set_regex(String typeLabel, Value.Type valueType, String regex) {
        if (!valueType.equals(Value.Type.STRING)) fail();
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        attributeType.setRegex(tx(), regex);
    }

    @Then("attribute\\( ?{type_label} ?) as\\( ?{value_type} ?) unset regex")
    public void attribute_type_as_value_type_unset_regex(String typeLabel, Value.Type valueType) {
        if (!valueType.equals(Value.Type.STRING)) fail();
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        attributeType.unsetRegex(tx());
    }

    @Then("attribute\\( ?{type_label} ?) as\\( ?{value_type} ?) get regex: {}")
    public void attribute_type_as_value_type_get_regex(String typeLabel, Value.Type valueType, String regex) {
        if (!valueType.equals(Value.Type.STRING)) fail();
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        assertEquals(regex, attributeType.getRegex(tx()));
    }

    @Then("attribute\\( ?{type_label} ?) as\\( ?{value_type} ?) does not have any regex")
    public void attribute_type_as_value_type_does_not_have_any_regex(String typeLabel, Value.Type valueType) {
        attribute_type_as_value_type_get_regex(typeLabel, valueType, null);
    }

    @Then("attribute\\( ?{type_label} ?) get owners, with annotations: {annotations}; contain:")
    public void attribute_type_get_owners_with_annotations_contain(String typeLabel, List<Annotation> annotations, List<String> ownerLabels) {
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        Set<String> actuals = attributeType.getOwners(tx(), set(annotations)).map(t -> t.getLabel().name()).collect(toSet());
        assertTrue(actuals.containsAll(ownerLabels));
    }

    @Then("attribute\\( ?{type_label} ?) get owners, with annotations: {annotations}; do not contain:")
    public void attribute_type_get_owners_with_annotations_do_not_contain(String typeLabel, List<Annotation> annotations, List<String> ownerLabels) {
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        Set<String> actuals = attributeType.getOwners(tx(), set(annotations)).map(t -> t.getLabel().name()).collect(toSet());
        for (String ownerLabel : ownerLabels) {
            assertFalse(actuals.contains(ownerLabel));
        }
    }

    @Then("attribute\\( ?{type_label} ?) get owners explicit, with annotations: {annotations}; contain:")
    public void attribute_type_get_owners_explicit_with_annotations_contain(String typeLabel, List<Annotation> annotations, List<String> ownerLabels) {
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        Set<String> actuals = attributeType.getOwners(tx(), set(annotations), EXPLICIT).map(t -> t.getLabel().name()).collect(toSet());
        assertTrue(actuals.containsAll(ownerLabels));
    }

    @Then("attribute\\( ?{type_label} ?) get owners explicit, with annotations: {annotations}; do not contain:")
    public void attribute_type_get_owners_explicit_with_annotations_do_not_contain(String typeLabel, List<Annotation> annotations, List<String> ownerLabels) {
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        Set<String> actuals = attributeType.getOwners(tx(), set(annotations), EXPLICIT).map(t -> t.getLabel().name()).collect(toSet());
        for (String ownerLabel : ownerLabels) {
            assertFalse(actuals.contains(ownerLabel));
        }
    }

    @Then("attribute\\( ?{type_label} ?) get owners contain:")
    public void attribute_type_get_owners_contain(String typeLabel, List<String> ownerLabels) {
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        Set<String> actuals = attributeType.getOwners(tx(), emptySet()).map(t -> t.getLabel().name()).collect(toSet());
        assertTrue(actuals.containsAll(ownerLabels));
    }

    @Then("attribute\\( ?{type_label} ?) get owners do not contain:")
    public void attribute_type_get_owners_do_not_contain(String typeLabel, List<String> ownerLabels) {
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        Set<String> actuals = attributeType.getOwners(tx(), emptySet()).map(t -> t.getLabel().name()).collect(toSet());
        for (String ownerLabel : ownerLabels) {
            assertFalse(actuals.contains(ownerLabel));
        }
    }

    @Then("attribute\\( ?{type_label} ?) get owners explicit contain:")
    public void attribute_type_get_owners_explicit_contain(String typeLabel, List<String> ownerLabels) {
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        Set<String> actuals = attributeType.getOwners(tx(), emptySet(), EXPLICIT).map(t -> t.getLabel().name()).collect(toSet());
        assertTrue(actuals.containsAll(ownerLabels));
    }

    @Then("attribute\\( ?{type_label} ?) get owners explicit do not contain:")
    public void attribute_type_get_owners_explicit_do_not_contain(String typeLabel, List<String> ownerLabels) {
        AttributeType attributeType = tx().concepts().getAttributeType(typeLabel);
        Set<String> actuals = attributeType.getOwners(tx(), emptySet(), EXPLICIT).map(t -> t.getLabel().name()).collect(toSet());
        for (String ownerLabel : ownerLabels) {
            assertFalse(actuals.contains(ownerLabel));
        }
    }
}

File: csharp/Test/Behaviour/Concept/Type/AttributeType/AttributeTypeSteps.cs
/*
 * Copyright (C) 2022 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using DataTable = Gherkin.Ast.DataTable;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Xunit;
using Xunit.Gherkin.Quick;
using static Vaticle.Typedb.Driver.Api.IConcept.Transitivity;

using Vaticle.Typedb.Driver;
using Vaticle.Typedb.Driver.Api;
using Vaticle.Typedb.Driver.Common;
using static Vaticle.Typedb.Driver.Api.IThingType;

namespace Vaticle.Typedb.Driver.Test.Behaviour
{
    public partial class BehaviourSteps
    {
        public IValue.ValueType GetValueType(string type)
        {
            switch (type)
            {
                case "long":
                    return IValue.ValueType.LONG;
                case "double":
                    return IValue.ValueType.DOUBLE;
                case "string":
                    return IValue.ValueType.STRING;
                case "boolean":
                    return IValue.ValueType.BOOL;
                case "datetime":
                    return IValue.ValueType.DATETIME;
                default:
                    throw new BehaviourTestException($"Unexpected value type {type}");
            }
        }

        [Given(@"put attribute type: ([a-zA-Z0-9-_]+), with value type: {}")]
        public void PutAttributeTypeWithValueType(string typeLabel, string valueTypeData)
        {
            IValue.ValueType valueType = GetValueType(valueTypeData);
            Tx.Concepts.PutAttributeType(typeLabel, valueType).Resolve();
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) get value type: {}")]
        public void AttributeTypeGetValueType(string typeLabel, string valueTypeData)
        {
            IValue.ValueType valueType = GetValueType(valueTypeData);
            Assert.Equal(
                valueType,
                Tx.Concepts.GetAttributeType(typeLabel).Resolve().ValueType);
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) get supertype value type: {}")]
        public void AttributeTypeGetSupertypeValueType(string typeLabel, string valueTypeData)
        {
            IValue.ValueType valueType = GetValueType(valueTypeData);
            IAttributeType supertype = Tx.Concepts
                .GetAttributeType(typeLabel).Resolve()
                .GetSupertype(Tx).Resolve()
                .AsAttributeType();

            Assert.Equal(valueType, supertype.ValueType);
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) as\({}\) get subtypes contain:")]
        public void AttributeTypeAsValueTypeGetSubtypesContain(
            string typeLabel, string valueTypeData, DataTable subLabelsData)
        {
            IValue.ValueType valueType = GetValueType(valueTypeData);
            var subLabels = Util.ParseDataTableToTypeList<string>(subLabelsData, val => val.ToString());

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            HashSet<string> actuals = attributeType
                .GetSubtypes(Tx, valueType)
                .Select(t => t.Label.Name)
                .ToHashSet();

            Assert.False(actuals.Except(subLabels).Any());
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) as\({}\) get subtypes do not contain:")]
        public void AttributeTypeAsValueTypeGetSubtypesDoNotContain(
        string typeLabel, string valueTypeData, DataTable subLabelsData)
        {
            IValue.ValueType valueType = GetValueType(valueTypeData);
            var subLabels = Util.ParseDataTableToTypeList<string>(subLabelsData, val => val.ToString());

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            HashSet<string> actuals = attributeType
                .GetSubtypes(Tx, valueType)
                .Select(t => t.Label.Name)
                .ToHashSet();

            foreach (string subLabel in subLabels)
            {
                Assert.False(actuals.Contains(subLabel));
            }
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) as\({}\) set regex: {}")]
        public void AttributeTypeAsValueTypeSetRegex(string typeLabel, string valueTypeData, string regex)
        {
            IValue.ValueType valueType = GetValueType(valueTypeData);
            Assert.True(valueType.Equals(IValue.ValueType.STRING));

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            attributeType.SetRegex(Tx, regex).Resolve();
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) as\({}\) unset regex")]
        public void AttributeTypeAsValueTypeUnsetRegex(string typeLabel, string valueTypeData)
        {
            IValue.ValueType valueType = GetValueType(valueTypeData);
            Assert.True(valueType.Equals(IValue.ValueType.STRING));

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            attributeType.UnsetRegex(Tx).Resolve();
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) as\({}\) get regex: {}")]
        public void AttributeTypeAsValueTypeGetRegex(string typeLabel, string valueTypeData, string regex)
        {
            IValue.ValueType valueType = GetValueType(valueTypeData);
            Assert.True(valueType.Equals(IValue.ValueType.STRING));

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            Assert.Equal(regex, attributeType.GetRegex(Tx).Resolve());
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) as\({}\) does not have any regex")]
        public void AttributeTypeAsValueTypeDoesNotHaveAnyRegex(string typeLabel, string valueTypeData)
        {
            AttributeTypeAsValueTypeGetRegex(typeLabel, valueTypeData, null);
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) get owners, with annotations: (\s*([\w\-_]+,\s*)*[\w\-_]*\s*); contain:")]
        public void AttributeTypeGetOwnersWithAnnotationsContain(
            string typeLabel, string annotationsData, DataTable ownerLabelsData)
        {
            List<Annotation> annotations = GetAnnotations(Util.ParseEnumerationToList(annotationsData));
        
            var ownerLabels = Util.ParseDataTableToTypeList<string>(ownerLabelsData, val => val.ToString());

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            HashSet<string> actuals = attributeType
                .GetOwners(Tx, annotations)
                .Select(t => t.Label.Name)
                .ToHashSet();

            Assert.False(actuals.Except(ownerLabels).Any());
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) get owners, with annotations: (\s*([\w\-_]+,\s*)*[\w\-_]*\s*); do not contain:")]
        public void AttributeTypeGetOwnersWithAnnotationsDoNotContain(
            string typeLabel, string annotationsData, DataTable ownerLabelsData)
        {
            List<Annotation> annotations = GetAnnotations(Util.ParseEnumerationToList(annotationsData));

            var ownerLabels = Util.ParseDataTableToTypeList<string>(ownerLabelsData, val => val.ToString());

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            HashSet<string> actuals = attributeType
                .GetOwners(Tx, annotations)
                .Select(t => t.Label.Name)
                .ToHashSet();

            foreach (string ownerLabel in ownerLabels)
            {
                Assert.False(actuals.Contains(ownerLabel));
            }
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) get owners explicit, with annotations: (\s*([\w\-_]+,\s*)*[\w\-_]*\s*); contain:")]
        public void AttributeTypeGetOwnersExplicitWithAnnotationsContain(
            string typeLabel, string annotationsData, DataTable ownerLabelsData)
        {
            List<Annotation> annotations = GetAnnotations(Util.ParseEnumerationToList(annotationsData));

            var ownerLabels = Util.ParseDataTableToTypeList<string>(ownerLabelsData, val => val.ToString());

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            HashSet<string> actuals = attributeType
                .GetOwners(Tx, annotations, EXPLICIT)
                .Select(t => t.Label.Name)
                .ToHashSet();

            Assert.False(actuals.Except(ownerLabels).Any());
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) get owners explicit, with annotations: (\s*([\w\-_]+,\s*)*[\w\-_]*\s*); do not contain:")]
        public void AttributeTypeGetOwnersExplicitWithAnnotationsDoNotContain(
            string typeLabel, string annotationsData, DataTable ownerLabelsData)
        {
            List<Annotation> annotations = GetAnnotations(Util.ParseEnumerationToList(annotationsData));

            var ownerLabels = Util.ParseDataTableToTypeList<string>(ownerLabelsData, val => val.ToString());

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            HashSet<string> actuals = attributeType
                .GetOwners(Tx, annotations, EXPLICIT)
                .Select(t => t.Label.Name)
                .ToHashSet();

            foreach (string ownerLabel in ownerLabels)
            {
                Assert.False(actuals.Contains(ownerLabel));
            }
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) get owners contain:")]
        public void AttributeTypeGetOwnersContain(string typeLabel, DataTable ownerLabelsData)
        {
            var ownerLabels = Util.ParseDataTableToTypeList<string>(ownerLabelsData, val => val.ToString());

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            HashSet<string> actuals = attributeType
                .GetOwners(Tx, new Annotation[0])
                .Select(t => t.Label.Name)
                .ToHashSet();

            Assert.False(actuals.Except(ownerLabels).Any());
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) get owners do not contain:")]
        public void AttributeTypeGetOwnersDoNotContain(string typeLabel, DataTable ownerLabelsData)
        {
            var ownerLabels = Util.ParseDataTableToTypeList<string>(ownerLabelsData, val => val.ToString());

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            HashSet<string> actuals = attributeType
                .GetOwners(Tx, new Annotation[0])
                .Select(t => t.Label.Name)
                .ToHashSet();

            foreach (string ownerLabel in ownerLabels)
            {
                Assert.False(actuals.Contains(ownerLabel));
            }
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) get owners explicit contain:")]
        public void AttributeTypeGetOwnersExplicitContain(string typeLabel, DataTable ownerLabelsData)
        {
            var ownerLabels = Util.ParseDataTableToTypeList<string>(ownerLabelsData, val => val.ToString());

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            HashSet<string> actuals = attributeType
                .GetOwners(Tx, new Annotation[0], EXPLICIT)
                .Select(t => t.Label.Name)
                .ToHashSet();

            Assert.False(actuals.Except(ownerLabels).Any());
        }

        [Then(@"attribute(\([a-zA-Z0-9-_]+\)) get owners explicit do not contain:")]
        public void AttributeTypeGetOwnersExplicitDoNotContain(string typeLabel, DataTable ownerLabelsData)
        {
            var ownerLabels = Util.ParseDataTableToTypeList<string>(ownerLabelsData, val => val.ToString());

            IAttributeType attributeType = Tx.Concepts.GetAttributeType(typeLabel).Resolve();
            HashSet<string> actuals = attributeType
                .GetOwners(Tx, new Annotation[0], EXPLICIT)
                .Select(t => t.Label.Name)
                .ToHashSet();

            foreach (string ownerLabel in ownerLabels)
            {
                Assert.False(actuals.Contains(ownerLabel));
            }
        }
    }
}
